src/main/scala/br/unb/cic/oberon/cfg/ControlFlowGraphBuilder.scala

```diff
@@ -5,8 +5,18 @@ import scalax.collection.mutable.Graph
 import scalax.collection.GraphEdge
 import scalax.collection.GraphPredef.EdgeAssoc
 
-import scala.+:
-import scala.collection.mutable.Stack
+abstract class StmtHolder[+A] {
+  def get: A
+}
+case object NoStmt extends StmtHolder[Nothing] {
+  def get: Nothing = throw new NoSuchElementException("NoStmt.get")
+}
+case class FullStmt(stmt: Statement) extends StmtHolder[Statement] {
+  def get: Statement = stmt
+}
+case class IgnoreInitStmt(stmt: Statement) extends StmtHolder[Statement] {
+  def get: Statement = stmt
+}
 
```

Comment:
could not understand the need for StmtHolder. Please, elaborate on that.

Reply:
ForStmt and RepeatUntilStmt need it, as they are loops that have sub-statements that go before themselves, I will call those pre-stmts.

To help illustrate why it is needed I will give an example, first without using it. Imagine the AST is composed of 3 statements, a simple statament 1, a ForStmt and another simple statement 2. The ForStmt has an Init Statement and an Inner Statement. We will assume Statements 1, 2, Init and Inner are simple just to make this explanation easier.

![AST view](https://user-images.githubusercontent.com/13670489/156188157-d2168249-2330-4b71-be47-c278cc0e2cf6.png)

When IntraProceduralGraphBuilder tries to create an edge from Statement 1 to FromStmt using addEdge it will do the following: addEdge will receive Statement 1 as "from" and ForStmt as "target", then it will create an edge from Statement 1(from) to the Init Statement, then it will create an edge from the Init Statement to the ForStmt(target).

![image](https://user-images.githubusercontent.com/13670489/156189750-a2dfe3a8-c622-47b2-af47-1ded154b21e6.png)

Indirect edge created, now we process Statement 1 to deal with its indirections. As it doesn't have any, we move on to creating an edge from ForStmt to Statement 2, quite simple.
 
![image](https://user-images.githubusercontent.com/13670489/156191924-b6679d1c-3b85-4901-a486-1d769e636cbe.png)

But at this point ForStmt will be processed as well. This will eventually call processStmtNode with ForStmt as "from", Inner as "target" and ForStmt as "end" (this is a loop, after all). An edge will be created from ForStmt to Inner.

![image](https://user-images.githubusercontent.com/13670489/156194452-985aa1f1-ccbb-4cdd-8a78-6f380f0c4068.png)

Now processStmtNode must also create an edge from "target" to "end", but how? If we simply use addEdge(target, end), Inner would be connected to Init, which is not intended.

![image](https://user-images.githubusercontent.com/13670489/156194733-68aec338-1a97-472d-8631-77c21d76ba56.png)

We also could create the last edge by directly creating it from "target" to "end", as was done before. That would achieve the right result in this example, but the function processStmtNode is also used to deal with IfElseStmts, connecting an edge directly from an IfElseStmts to the "end" would create problems if "end" was a ForStmt:

![image](https://user-images.githubusercontent.com/13670489/156195933-724e4a7b-134b-4a28-aea0-ed787bc5ffb7.png)

How to solve this? I needed a way to differentiate whether the statement should be considered in full, including the pre-stmts beforehand, or if the pre-stmts should be disregarded.

Another way to achieve the right result in theses cases would be to **create a clone of processStmtNode to deal with loops specifically**. But if "target" is a StmtSequence processStmtNode delegates to createControlFlowGraph, so the problem persists and we would have to create a clone of createControlFlowGraph to deal with loops as well... seems to be a bad move.

tl;dr: What I ended up doing was: create a case class that will hold the information of WHERE the graph is supposed to go back to. I alse gave it a "get" function that mimics the behaviour of Option.get so that the cases where "get" is used don't change.

Is there a better solution to this? I guess I could create a simpler case class to encode the two valued cases and wrap it in an Option to get the NoStmt case as well. But I am not sure if that is better.

Reply:
I do not think this is necessary. I will take a look at the test cases to confirm that `for` and `repeat until` statements do require such special treatment. IMO, this part of the PR is adding too much unnecessary complexity.

Reply:
I understand your point. But there is a clever solution for building a control flow graph. I think that it is time to experiment with such an approach. Please, take a look at the book Principles of Program Analysis, sections 1.1 and 2.1. This book is freely available from Periódicos CAPES. 

Another option, which is our goal for this semester, is to convert every `repeat until`, `for` loop, and that like into `while` statements. 

Please, let me know which path are you planning to follow first.

Reply:
anyway, we already have a transformation that converts Oberon into Oberon Core.

Reply:
I can find some time tomorrow to improve the control flow graph, considering the recommendations from Principles of Program Analysis. Please, let me know if you prefer this alternative.

Reply:
I think that would be great! Would you like me to create a new PR only with the addition of the Ids to statements? I think that alone fixes one of the tests (but it may break others, i haven't checked yet)

Reply:
Hi @delanoerhardt . I would love that. Could you please consider implementing a trait that has a label function (your Id) and then make Stmt mix with such a trait in the control flow graph module? In this way we could isolate this change better.

Many, many thanks for the discussion.

Reply:
It seems like it is impossible to isolate the changes here.

We need to change the behaviour of either "==" comparison, hashCode or some other function that has a role in defining equality.

I tried using implicit classes, but it seems that those can't be used for overriding methods already defined, as is the case with the mentioned methods.

If you have any directions on how to do this I will look into it, otherwise this PR should be closed.

Reply:
I thought of a way of fixing this just before sleeping lol.

Using implicit classes wasn't even needed, we have access to the SimpleNode class on this module.

I will create another PR with the fix